src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -861,6 +861,84 @@ void StorageReplicatedMergeTree::createReplica(const StorageMetadataPtr & metada
 
     LOG_DEBUG(log, "Creating replica {}", replica_path);
 
+    const String local_metadata = ReplicatedMergeTreeTableMetadata(*this, metadata_snapshot).toString();
+    const String local_columns = metadata_snapshot->getColumns().toString();
+    const String local_metadata_version = toString(metadata_snapshot->getMetadataVersion());
+
+    /// It is possible for the replica to fail after creating ZK nodes without saving local metadata.
+    /// Because of that we need to check whether the replica exists and is newly created.
+    /// For this we check that all nodes exist, the metadata of the table is the same, and other nodes are not modified.
+
+    std::vector<String> paths_exists = {
+        replica_path,
+        replica_path + "/host",
+        replica_path + "/log_pointer",
+        replica_path + "/queue",
+        replica_path + "/parts",
+        replica_path + "/flags",
+        replica_path + "/is_lost",
+        replica_path + "/metadata",
+        replica_path + "/columns",
+        replica_path + "/metadata_version",
+        replica_path + "/mutation_pointer",
+        replica_path + "/min_unprocessed_insert_time",
+        replica_path + "/max_processed_insert_time",
+        replica_path + "/mutation_pointer"
+    };
+
+    auto response_exists = zookeeper->tryGet(paths_exists);
+    size_t response_num = 0;
+
+    if (response_exists[response_num++].error == Coordination::Error::ZOK)
+    {
+        bool all_nodes_exist = true;
+
+        for (size_t i = 0; i < response_exists.size(); ++i)
+        {
+            if (response_exists[i].error != Coordination::Error::ZOK)
+            {
+                all_nodes_exist = false;
+                break;
+            }
+        }
+
+        if (all_nodes_exist)
+        {
+            const auto & zk_host                        = response_exists[response_num++].data;
+            const auto & zk_log_pointer                 = response_exists[response_num++].data;
+            const auto & zk_queue                       = response_exists[response_num++].data;
+            const auto & zk_parts                       = response_exists[response_num++].data;
+            const auto & zk_flags                       = response_exists[response_num++].data;
+            const auto & zk_is_lost                     = response_exists[response_num++].data;
+            const auto & zk_metadata                    = response_exists[response_num++].data;
+            const auto & zk_columns                     = response_exists[response_num++].data;
+            const auto & zk_metadata_version            = response_exists[response_num++].data;
+            const auto & zk_min_unprocessed_insert_time = response_exists[response_num++].data;
+            const auto & zk_max_processed_insert_time   = response_exists[response_num++].data;
+            const auto & zk_mutation_pointer            = response_exists[response_num++].data;
+
+            if (zk_host.empty() &&
```

Comment:
Will this code actually distinguish the situation when the replica was not created completely because of expired session from the situation when a replica with this name already exists, but is managed by other clickhouse-server? Imagine that we are executing `CREATE ... ON CLUSTER ... ENGINE=ReplicatedMergeTree('/some/path', '{replica}')`, but the cluster is misconfigured, and two replicas have the same value of `{replica}` macro in the config.

Reply:
There can be a case where two replicas simultaneously create the table with the same `{replica}` value and are both ~in this function at the same time, then it seems like everything will go south. The problem is that the replicas cannot really be distinguished until the storage is started. I added the `creator_uuid` node, let me know if you have better ideas.

Reply:
Consider also adding the (qualified) table name somewhere (it's okay to simply concatenate it with the uuid and write into creator_uuid), just in case someone tries to create two tables with the same replica_path on one server (they will have the same host UUID in this case)

Reply:
> Consider also adding the (qualified) table name

Or table's UUID